### ISSUE TITLE
Trying to Edit Frozen String - 0.44.0 Release

### ISSUE THREAD

**Issue by adstage-david** [excon/excon#468]
Looks like something changed in the latest release that started breaking our staging environment. 

We have a constant defined at boot to represent an endoint that we use Excon to connect to, all of a sudden Airbrake has been flooded with:

```
RuntimeError: can't modify frozen String
/vendor/bundle/ruby/2.0.0/gems/excon-0.44.0/lib/excon/utils.rb:79 in "force_encoding"
/vendor/bundle/ruby/2.0.0/gems/excon-0.44.0/lib/excon/utils.rb:79 in "unescape_form"
/vendor/bundle/ruby/2.0.0/gems/excon-0.44.0/lib/excon/connection.rb:74 in "initialize"
/vendor/bundle/ruby/2.0.0/gems/excon-0.44.0/lib/excon.rb:126 in "new"
/vendor/bundle/ruby/2.0.0/gems/excon-0.44.0/lib/excon.rb:126 in "new"
/app/models/metrics_api/connections/api_adapter.rb:24 in "build_connection!"
/app/models/metrics_api/connections/api_adapter.rb:18 in "initialize"
```

Should be pretty easy to work around, for now we've just locked to 0.43 - but wanted to bring up in case this is not intentional.

**Comment by geemus** [comment]
I haven't seen broader issues here just yet. I think it might come from either a username or password value being frozen? (or perhaps if the user/pass are parsed out of basic auth a URL which is frozen). Does that sound like it could match your case? Perhaps we don't actually need to force_encode here, but we didn't think it would be harmful (at least in most cases).

**Comment by adstage-david** [comment]
The root url, the user and the password are all frozen, which is something I have't really thought about, but apparently Ruby does for all ENV vars. 

They get passed into our Heroku app with ENV['METRICS_ROOT'], ENV['METRICS_USER'] and ENV['METRICS_PASSWORD']

**Comment by jrafanie** [comment]
Looks like this came from #457.  I'm not sure if this was an oversight or intentional.

If the incoming strings are coming from the user/client code, we should dup them if we need to use force_encoding as force_encoding modifies them in place.

```
ruby -e '"a".freeze.force_encoding("US-ASCII")'
-e:1:in `force_encoding': can't modify frozen String (RuntimeError)
```

Note, the gsub! and force_encoding will both mutate the receiver so we should dup the string in each of these methods or in the callers:

```ruby
    # Splits a header value +str+ according to HTTP specification.
    def split_header_value(str)
      return [] if str.nil?
      str = str.strip
      str.force_encoding('BINARY') if FORCE_ENC
      str.scan(%r'\G((?:"(?:\\.|[^"])+?"|[^",]+)+)
                    (?:,\s*|\Z)'xn).flatten
    end

    # Escapes HTTP reserved and unwise characters in +str+
    def escape_uri(str)
      str.force_encoding('BINARY') if FORCE_ENC
      str.gsub(UNESCAPED) { "%%%02X" % $1[0].ord }
    end

    # Unescapes HTTP reserved and unwise characters in +str+
    def unescape_uri(str)
      str.force_encoding('BINARY') if FORCE_ENC
      str.gsub(ESCAPED) { $1.hex.chr }
    end

    # Unescape form encoded values in +str+
    def unescape_form(str)
      str.force_encoding('BINARY') if FORCE_ENC
      str.gsub!(/\+/, ' ')
      str.gsub(ESCAPED) { $1.hex.chr }
    end
```

**Comment by jrafanie** [comment]
cc @rusikf @dpiddy

**Comment by rusikf** [comment]
Yes, it is my fault because i don't tested it with frozen args, fixed it in #469

**Comment by jrafanie** [comment]
Thanks, @rusikf, now we have a test to prevent future regressions.

**Comment by geemus** [comment]
Thanks!

**Comment by geemus** [comment]
Should be fixed in 0.44.1
